### ISSUE TITLE
Suggestion for Fastify 5: Required param to .send

### ISSUE THREAD

**Issue by JonRC** [fastify/fastify#5534]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

I propose to make the param of the [`send`](https://github.com/fastify/fastify/blob/main/types/reply.d.ts#L53) method 'required' in the type declaration.
This way, if you do not use any [`RouteGenericInterface`](https://github.com/fastify/fastify/blob/main/types/route.d.ts#L22) or [`TypeProvider`](https://github.com/fastify/fastify/blob/main/types/type-provider.d.ts#L9), you will be free to pass anything to `send`, including nothing.
But, if you use any `RouteGenericInterface` or `TypeProvider`, the type checker will be more strict. You will not be able to reply `undefined` for a required reply's payload specified through `TypesProvider` or `RouteGenericInterface`.

### Motivation

My motivation was this issue https://github.com/turkerdev/fastify-type-provider-zod/issues/88

In this issue, the user asks about a more restricted type checking for replies.

### Example

```TS
const server = fastify()

interface ReplyHttpCodes {
  Reply: {
    201: string
    204: undefined
  }
}

server.get<ReplyHttpCodes>('/', async function handler (request, reply) {
  reply.status(201).send('some string') // No Error
  reply.status(204).send() // No Error
  reply.status(204).send(undefined) //No Error

  reply.status(201).send() // TS Error: Expected 1 arguments, but got 0 (My PR introduced this checking)
  reply.status(201).send(123) // TS Error: Argument of type 'number' is not assignable to parameter of type 'string'
  reply.status(204).send('543534') //TS Error: Argument of type 'string' is not assignable to parameter of type 'void' 
})

```

**Comment by JonRC** [comment]
I just created a draft PR if anyone wants to take a look at my idea.
https://github.com/fastify/fastify/pull/5535

**Comment by jsumners** [comment]
`reply.status(201).send()` is not an error.

**Comment by JonRC** [comment]
> `reply.status(201).send()` is not an error.

You're right, it's not an error itself. But when I specify the response type as `string` I suppose that an `undefined` or empty response wouldn't be acceptable.

I just updated the example, is it more clear?

**Comment by JonRC** [comment]
> `reply.status(201).send()` is not an error.

@jsumners 
You're right. Considering the example, there's no type declaration for the 201 code, so you're free to pass anything (or nothing as it's common for 201 status).

My point here is: if you specified that the reply type is `string`, `undefined` wouldn't be acceptable.

**Comment by jsumners** [comment]
I don't participate with TypeScript. I do not consider `reply.status(<any_code>).send()` to be an error.

**Comment by JonRC** [comment]
In terms of JS and HTTP, I agree with you.
But from a typescript point of view, it wouldn't be acceptable an empty response where a `string` _(or any type)_ is required.

@jsumners do you know who maintains the typescript part so they can contribute to this discussion?

**Comment by jsumners** [comment]
@fastify/typescript

**Comment by mcollina** [comment]
I don't think the implementation is correct, as returning `undefined` response is indeed valid and supports.

However, the problem is real. I think a ternary could help in this case. Essentially we need to enable `void` if a type for the response was not specified/inferred.

**Comment by JonRC** [comment]
@mcollina 
You're right. We need another solution for it. Give me some days to send another suggestion.

**Comment by JonRC** [comment]
Hey @mcollina and @kibertoad

Now, it's not possible to pass `void` to `send` when non `undefined` type is specified/inferred but it's still possible to pass `void` or `undefined` when `undefined` is specified/inferred.
Take a look at the examples and the PR.

```TS
const server = fastify()

interface ReplyHttpCodes {
  Reply: {
    201: string
    204: undefined
  }
}

server.get<ReplyHttpCodes>('/', async function handler (request, reply) {
  reply.status(201).send('some string') // No Error
  reply.status(204).send() // No Error
  reply.status(204).send(undefined) //No Error

  reply.status(201).send() // TS Error: Expected 1 arguments, but got 0 (My PR introduced this checking)
  reply.status(201).send(123) // TS Error: Argument of type 'number' is not assignable to parameter of type 'string'
  reply.status(204).send('543534') //TS Error: Argument of type 'string' is not assignable to parameter of type 'void' 
})
```

**Comment by johnkmzhou** [comment]
So the issue I've noticed with this is that Typebox, and relatedly JSON Schema, are unable to represent "no payload". The best Typebox can do is `Type.Null()` which due to this change now requires `reply.status(204).send(null)`. Before, you could get away just `reply.status(204).send()`.

**Comment by cmcnicholas** [comment]
> ch d

You can use `Type.Never()` and it tricks it, but I still don't feel comfortable using it as to me that indicates 204 => never
